From dc6801216c9c34e762634c81e91ba22ed9e70366 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Mon, 26 Mar 2018 15:19:12 +0800
Subject: [PATCH 3590/5242] MLK-17931-2 gpu: imx: dpu: common: Differentiate
 DPUs w/wo DPR fixups

commit  300c1e250bef42f2273a7d64e70fe4e9cc774f5c from
https://source.codeaurora.org/external/imx/linux-imx.git

DPR may prefetch linear frames and resolve tile frames for DPU.
New i.MX8QXP SoCs have DPR fixups in silicon.
This patch differentiates DPUs with or without DPR fixups
via OF device ID's data.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-common.c |   43 +++++++++++++++++++++++++++++++++++---
 drivers/gpu/imx/dpu/dpu-prv.h    |    1 +
 2 files changed, 41 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/imx/dpu/dpu-common.c b/drivers/gpu/imx/dpu/dpu-common.c
index 1b621ee..3d66ddf 100644
--- a/drivers/gpu/imx/dpu/dpu-common.c
+++ b/drivers/gpu/imx/dpu/dpu-common.c
@@ -542,6 +542,7 @@ static inline void dpu_cm_write(struct dpu_soc *dpu, u32 value,
 	.plane_src_na_mask = 0xffffff80,
 	.has_capture = true,
 	.has_prefetch = false,
+	.has_prefetch_fixup = false,
 	.pixel_link_quirks = false,
 	.pixel_link_nhvsync = false,
 	.version = DPU_V1,
@@ -570,15 +571,51 @@ static inline void dpu_cm_write(struct dpu_soc *dpu, u32 value,
 	.plane_src_na_mask = 0xffffffe2,
 	.has_capture = false,
 	.has_prefetch = true,
+	.has_prefetch_fixup = false,
+	.pixel_link_quirks = true,
+	.pixel_link_nhvsync = true,
+	.version = DPU_V2,
+};
+
+static const struct dpu_devtype dpu_type_v2_with_prefetch_fixup = {
+	.cm_ofs = 0x0,
+	.cfs = &cfs_v2,
+	.decs = &decs_v2,
+	.eds = &eds_v2,
+	.fds = &fds_v2,
+	.fes = &fes_v2,
+	.fgs = &fgs_v2,
+	.fls = &fls_v2,
+	.fws = &fws_v2,
+	.hss = &hss_v2,
+	.lbs = &lbs_v2,
+	.tcons = &tcons_v2,
+	.vss = &vss_v2,
+	.cm_reg_ofs = &cm_reg_ofs_v2,
+	.intsteer_map = intsteer_map_v2,
+	.intsteer_map_size = ARRAY_SIZE(intsteer_map_v2),
+	.unused_irq = unused_irq_v2,
+	.sw2hw_irq_map = sw2hw_irq_map_v2,
+	.sw2hw_block_id_map = sw2hw_block_id_map_v2,
+	.plane_src_na_mask = 0xffffffe2,
+	.has_capture = false,
+	.has_prefetch = true,
+	.has_prefetch_fixup = true,
 	.pixel_link_quirks = true,
 	.pixel_link_nhvsync = true,
 	.version = DPU_V2,
 };
 
 static const struct of_device_id dpu_dt_ids[] = {
-	{ .compatible = "fsl,imx8qm-dpu", .data = &dpu_type_v2, },
-	{ .compatible = "fsl,imx8qxp-dpu", .data = &dpu_type_v2, },
-	{ /* sentinel */ }
+	{
+		.compatible = "fsl,imx8qm-dpu",
+		.data = &dpu_type_v2,
+	}, {
+		.compatible = "fsl,imx8qxp-dpu",
+		.data = &dpu_type_v2_with_prefetch_fixup,
+	}, {
+		/* sentinel */
+	}
 };
 MODULE_DEVICE_TABLE(of, dpu_dt_ids);
 
diff --git a/drivers/gpu/imx/dpu/dpu-prv.h b/drivers/gpu/imx/dpu/dpu-prv.h
index b1305de..aa22c19 100644
--- a/drivers/gpu/imx/dpu/dpu-prv.h
+++ b/drivers/gpu/imx/dpu/dpu-prv.h
@@ -206,6 +206,7 @@ struct dpu_devtype {
 	const u32 plane_src_na_mask;
 	bool has_capture;
 	bool has_prefetch;
+	bool has_prefetch_fixup;
 	bool pixel_link_quirks;
 	bool pixel_link_nhvsync;	/* HSYNC and VSYNC high active */
 	unsigned int version;
-- 
1.7.9.5

